222 pixel inflation

JulietteFrancovich opened the pull request.

JulietteFrancovich commented:
@psomhorst this branch now contains the feature pixel inflations and the parameter TIV since they are interdependent. I have written tests for pixel inflations. Could you review the code and give your suggestions for pixel inflations and the tests? Once we are happy with that I will write the tests for parameter TIV and request review again for the entire pull request. Thanks!

psomhorst commented on tests/test_pixel_inflation.py:
```diff
@@ -0,0 +1,306 @@
+                            assert math.isclose(value_at_time, 1, abs_tol=0.01)
+
+
+def test_with_data(draeger1: Sequence, draeger2: Sequence, timpel1: Sequence, pytestconfig: pytest.Config):
```
This last test is **very** slow and will slow down continuous integration on the server. All tests are run on every push on a pull request, on every merge, et cetera. 

I think it's important to test whether the overal structure works on real data, but maybe it's better to only one dataset, or just a small part (e.g. 20 breaths) of a larger dataset.

psomhorst replied:
Tests are sufficiently fast now, I think.

psomhorst reviewed (commented):
I think this is a great PR with a lot of work and thought in it. There are some things I would improve, but overall it seems almost good to go.

The algorithm is quite slow. I would like to look into trying to speed it up. I have a suggestion on how to prevent iterating over all pixels. Maybe this will help in the performance. 

Please make sure the code is formatted using "ruff < 0.6". I will update the main branch to use the last version of ruff before 0.6 (currently 0.5.7). 

I have only reviewed pixel inflation and its test. I have not reviewed the TIV parameters class or the notebooks.

JulietteFrancovich commented on eitprocessing/features/pixel_inflation.py:
```diff
@@ -0,0 +1,173 @@
+            name="Pixel in- and deflation timing as determined by PixelInflation",
+            unit=None,
+            category="breath",
+            intervals=[
```
I agree that this does not seem logical. The intervals now run from middle to middle of the continuous data breath which is the interval in which we detect the start of the pixel inflation. It might make more sense to just use the start and end time of the pixel inflation as interval? To me it does make sense to store it as intervaldata just as breaths. But maybe you have a different suggestion?

psomhorst replied:
I would certainly not store it as the interval from middle to middle. If it is stored as intervaldata, I would store it with the same intervals of the continuous data Breaths each inflation is associated with. In that way, when you slice, you actually keep the pixel inflations that belong to all breaths that you select.

If we store them without any timing data, they lose their slice ability, which means you have to filter them yourself. 

For now, let's keep it as IntervalData, but set the interval as the start/end of the breath the pixel inflation is associated with. If it requires something different, it's probably a new type of storage (which I'm not keen on introducing).

JulietteFrancovich replied:
If the interval is set from start to end of a normal breath and the pixel inflation is slightly delayed, wouldn't that make you miss the end of the pixel inflation in the interval?

psomhorst replied:
Yes. However, is this a problem? If I want to select 10 breaths, I want 10 inflations per pixel. If a pixel is early or late, and you include everything that overlaps even 1 timeframe, you get 11 inflations for many/most pixels. This does not match the number of breaths.

I think for now this is the best we can do. Let's see how it works when using it, and whether we run into issues.

JulietteFrancovich commented on eitprocessing/features/pixel_inflation.py:
```diff
@@ -0,0 +1,173 @@
+                for i in range(len(indices_breath_middles) - 1)
+            ],
+            values=pixel_inflations,
+            parameters={},
```
Yes, I think that makes sense

JulietteFrancovich commented:
@psomhorst I have addressed your comments and tried to incorporate your suggestions (see comments). Please let me know if you have further comments/ concerns. I will start working on tests for parameter TIV so we can review that next.

psomhorst commented on eitprocessing/features/pixel_inflation.py:
```diff
@@ -0,0 +1,173 @@
+        pixel_inflations = np.full((len(breaths), rows, cols), None)
+
+        for row in range(rows):
+            for col in range(cols):
```
I tried some stuff, but it seems impossible to do it without iteration. I did try to move determining the starts/end of the pixels breaths outside the loop (which is possible), but it is not faster.

psomhorst commented on notebooks/test_pixel_inflation.ipynb:
If you add a notebook, it serves an example of how to use the class. Could you clean this up and add some explanation?

JulietteFrancovich replied:
I have cleaned this notebook and the one for tiv as well and added explanation.

psomhorst reviewed (commented):
I have some additional comments about the PixelInflation tests and notebook.

psomhorst commented on eitprocessing/parameters/tidal_impedance_variation.py:
```diff
@@ -0,0 +1,205 @@
+            )
+            breath_data = global_breaths.values
+
+        number_of_breaths = breath_data.shape[0] if tiv_timing == "pixel" else len(breath_data)
```
Shouldn't the number of breaths always be the same? The number of pixel inflations is the same as the number of breaths, right?

JulietteFrancovich replied:
Yes they should always give the same result but depending on whether the tiv_timing is pixel or continuous the breath_data is either a list of which the length is the number of breaths or a (x, 32, 32) array of which the first element is the number of breaths. So I implemented this to be able to find the number of breaths regardless of the tiv_timing.

psomhorst replied:
Doesn't `len(breath_data)` return `shape[0]` if it's a multidimensional array?

```suggestion
        number_of_breaths = len(breath_data)
```

JulietteFrancovich replied:
Apparently yes 😅 I have updated it

JulietteFrancovich commented on eitprocessing/parameters/tidal_impedance_variation.py:
```diff
@@ -0,0 +1,205 @@
+
+@dataclass
+class TIV(ParameterCalculation):
+    """Compute the tidal impedance variation (TIV) per breath."""
```
When we first started working on pixel inflations we discussed that we shouldn't call it breaths since pixels do not have independent breaths but are always associated with breaths in the continuous data. But I agree that it can cause confusion. Would you suggest changing it to pixel breaths or rather explaining the nomenclature and keeping it consistent?

psomhorst replied:
I still think breath is not the best name, but then again, neither is inflation, since impedance changes are not always directly due to inflation, but can be due to e.g. movement. Maybe it's easiest to just use breath everywhere, and make a small footnote that impedance changes are not necessarily just due to inflation.

JulietteFrancovich replied:
I agree, for consistency I have changed it to pixel breaths.

JulietteFrancovich commented on eitprocessing/parameters/tidal_impedance_variation.py:
```diff
@@ -0,0 +1,205 @@
+        number_of_breaths = breath_data.shape[0] if tiv_timing == "pixel" else len(breath_data)
+        tiv_values_array = np.empty((number_of_breaths, rows, cols), dtype=object)
+
+        for i in range(number_of_breaths):
+            for row in range(rows):
+                for col in range(cols):
+                    if tiv_timing == "pixel" and not breath_data[i, row, col]:
+                        tiv_values_array[i, row, col] = np.nan
+                    else:
+                        time_series = data[:, row, col]
+                        tiv_values = self._calculate_tiv_values(
+                            time_series,
+                            eit_data.time,
+                            breath_data[:, row, col] if tiv_timing == "pixel" else breath_data,
+                            tiv_method,
+                            tiv_timing,
+                        )
+                        tiv_values_array[i, row, col] = tiv_values[i]
```
That is interesting.. I have changed it according to your suggestion and added a way to handle if all breaths are None in the _calculate_tiv_values methods.

JulietteFrancovich commented on tests/test_pixel_inflation.py:
```diff
@@ -0,0 +1,315 @@
+        else:
+            for row in range(2):
+                for col in range(2):
+                    if mean == 0:
```
During last review round we changed from checking whether the mean was 0 to if np.any(pixel_impedance[:, row, col] > 0). Therefore I removed mean == 0 from this test and created a separate test to see if it handles correctly if a pixel has zero impedance.

JulietteFrancovich commented on tests/test_parameter_tiv.py:
```diff
@@ -0,0 +1,400 @@
+    """Class to create Mock Sequence for running tests."""
+
+    def __init__(self, mock_eit_data: MockEITData, mock_continuous_data: MockContinuousData):
+        self.eit_data = mock_eit_data
```
If I change it to a data collection I need to initialise all the things that are needed for EITData but in the way I wrote it it works for the functions I want to test.

psomhorst replied:
It may work, but it is not clear nor correct. The attributes of the mocked sequence are not the attributes of normal sequences. That does not convince me the tests represent the actual code.

psomhorst commented on eitprocessing/features/pixel_inflation.py:
```diff
@@ -0,0 +1,203 @@
+            unit=None,
+            category="breath",
+            intervals=intervals,
+            values=pixel_inflations,
```
We can't ship this PR without fixing slicing for this specific result set. I think a separate PR is easier to manage. I made an issue (#304).

psomhorst replied:
In the meantime, what happens if you save the results as a list of 2D-arrays, rather than a 3D array?

psomhorst reviewed (commented):
I have some more comments.

JulietteFrancovich commented on tests/test_parameter_tiv.py:
```diff
@@ -0,0 +1,400 @@
+def test_compute_continuous_parameter_not_implemented_method(mock_continuous_data: MockContinuousData):
+    """Test that compute_continuous_parameter raises NotImplementedError for unsupported methods."""
+    tiv = TIV(method="unsupported_method")
+    with pytest.raises(NotImplementedError, match="Method unsupported_method is not implemented."):
```
I have moved the NotImplementedError to the initialisation and adjusted the test accordingly.

JulietteFrancovich commented on tests/test_parameter_tiv.py:
```diff
@@ -0,0 +1,400 @@
+timpel_file = data_directory / "Timpel_Test.txt"
+
+
+class MockEITData(EITData):
```
yes that makes sense! I have updated the tests accordingly.

psomhorst commented on eitprocessing/features/pixel_breath.py:
```diff
@@ -0,0 +1,208 @@
+
+@dataclass
+class PixelBreath:
+    """Algorithm for detecting timing of pixel breath and deflation in pixel impedance data.
```
You probably replace "inflation" with breath everywhere, resulting is a weird sentence here. Please check the documentation.

psomhorst commented on eitprocessing/parameters/tidal_impedance_variation.py:
```diff
@@ -0,0 +1,194 @@
+    def compute_parameter(
+        self,
+        data: ContinuousData | EITData,
+    ) -> str:
```
A single dispatch-method should not have a return type. This function never returns anything. The overloaded versions (compute_continuous_parameter and compute_pixel_parameter) should have proper return types.

JulietteFrancovich commented on eitprocessing/features/pixel_breath.py:
```diff
@@ -0,0 +1,208 @@
+            tiv_timing="continuous",
+        )
+
+        mean_tiv_pixel = np.nanmean(pixel_tiv_with_continuous_data_timing, axis=0)
```
I have integrated a test to only compute the nanmean for slices that are not entirely NaN.

JulietteFrancovich commented:
@psomhorst I have addressed your comments and changed the TIV to be stored as SparseData. I have adjusted the pixel breath accordingly (since this uses TIV) and have added new tests to check the storage of TIV. Let me know if you have any further comments :)

psomhorst reviewed (approved):
Fantastic! Great effort, thanks for the change to SparseData.

I took the liberty to rebase the branch onto develop, so you can go ahead and merge it.

JulietteFrancovich commented:
@psomhorst Thank you! I have merged and deleted the branch.
